st2reactor/st2reactor/rules/matcher.py

```diff
@@ -26,9 +27,40 @@ def __init__(self, trigger_instance, trigger, rules):
         self.rules = rules
 
     def get_matching_rules(self):
+        first_pass, second_pass = self._split_rules_into_passes()
+        # first pass
         rule_filters = [RuleFilter(self.trigger_instance, self.trigger, rule)
-                        for rule in self.rules]
+                        for rule in first_pass]
         matched_rules = [rule_filter.rule for rule_filter in rule_filters if rule_filter.filter()]
+        LOG.debug('[1st_pass] %d rule(s) found to enforce for %s.', len(matched_rules),
+                  self.trigger['name'])
+        # second pass
+        rule_filters = [SecondPassRuleFilter(self.trigger_instance, self.trigger, rule,
+                                             matched_rules)
+                        for rule in second_pass]
+        matched_in_second_pass = [rule_filter.rule for rule_filter in rule_filters
+                                  if rule_filter.filter()]
+        LOG.debug('[2nd_pass] %d rule(s) found to enforce for %s.', len(matched_in_second_pass),
+                  self.trigger['name'])
+        matched_rules.extend(matched_in_second_pass)
         LOG.info('%d rule(s) found to enforce for %s.', len(matched_rules),
                  self.trigger['name'])
         return matched_rules
+
+    def _split_rules_into_passes(self):
+        """
+        Splits the rules in the Matcher into first_pass and second_pass collections.
+
+        Since the
+        """
+        first_pass = []
+        second_pass = []
+        for rule in self.rules:
+            if self._is_first_pass_rule(rule):
+                first_pass.append(rule)
+            else:
+                second_pass.append(rule)
+        return first_pass, second_pass
+
+    def _is_first_pass_rule(self, rule):
+        return rule.type['ref'] != RULE_TYPE_BACKSTOP
```

Comment:
Won't this fail for all the existing rules in the DB since we only populate `type` attribute if the database object contains this attribute (https://github.com/StackStorm/st2/pull/1922/files#diff-d8db8c8c2b0e88de1bec747fd4b3d0a1R243)?